### PR TITLE
Add methods to set data, flags, and nsamples

### LINKED ISSUE #886: Consider making `put_*` methods matching `get_*` methods
A request form James, needs more thinking.

### PR DESCRIPTION
**Pull request by plaplant**
## Description

This PR adds the ability for the user to set data, flags, and nsamples arrays for UVDat objects. This is essentially the inverse of the `UVData.get_data()` et al. methods. By providing an appropriately sized array and the corresponding keys that would fetch data, the user is able to overwrite the data in the arrays without having to index into the datasets by hand.

## Motivation and Context

This PR leverages the fancy dataset indexing that was originally developed for performing optimized reading of UVH5 data. It has been made more general to support setting data in arrays, but the reading portion of the methods should work the same as before.

Still to do:
* add more test coverage
* add examples to the tutorial
* update the CHANGELOG

Closes #886.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/CHANGELOG.md).

### CODE REVIEW

File: pyuvdata/uvdata/tests/test_uvdata.py [hunk 1]
```diff
@@ -10948,4 +10948,324 @@ def test_eq_allowed_failures_filename_string(bda_test_file, capsys):
...
+    Test an error is raised when a key has too many values.
+    """
+    uv = hera_uvh5
+
+    ant1 = np.unique(uv.antenna_numbers)[0]
+    ant2 = np.unique(uv.antenna_numbers)[1]
+    data = uv.get_data(ant1, ant2, squeeze="none", force_copy=True)
+    with pytest.raises(ValueError) as cm:
+        uv.set_data(data, (ant1, ant2, "xx", "foo"))
+    assert str(cm.value).startswith("no more than 3 key values can be passed")
```
Comment (bhazelton) [inline pyuvdata/uvdata/tests/test_uvdata.py:11115]: you could just use the `match` keyword in the pytest.raises call (here and below).

**Review by bhazelton — COMMENTED** [review]
It seems like there's a lot of copy and pasted logic among the different set methods. Could we move some of that into a helper function to reduce that?

File: pyuvdata/uvdata/uvdata.py [hunk 2]
```diff
@@ -3437,6 +3443,231 @@ def get_ENU_antpos(self, center=False, pick_data_ants=False):
...
+        """
+        Set the nsamples array to some values provided by the user.
+
+        Parameters
+        ----------
+        nsamples : ndarray of float
+            The nsamples to overwrite into the nsample_array. Must be the same
+            shape as the target indices.
+        key1, key2, key3 : int or tuple of ints
+            Identifier of which nsamples to get, can be passed as 1, 2, or 3
```
Comment (bhazelton) [inline pyuvdata/uvdata/uvdata.py:3634]: get -> set

File: docs/tutorial.rst [hunk 3]
```diff
@@ -411,6 +411,23 @@ g) Quick access to file attributes of a UV* object (UVData, UVCal, UVBeam)
   >>> # Load object to instance name "uv" and will remain in interpreter
   pyuvdata_inspect.py -i <uv*_file>
 
+h) Update data arrays in place for UVData
```
Comment (bhazelton) [inline docs/tutorial.rst:414]: I think maybe this should be moved to between the existing d & e items, it should certainly come before the existing g item. 

Maybe we should also update the top-level text under this whole section ("UVData: Quick data access"), it's pretty terse at the moment.

**Comment by bhazelton** [comment]
I think we also need to update the changelog on this PR.

File: pyuvdata/uvdata/tests/test_uvdata.py (hunk 1 above)
Comment (plaplant) [reply pyuvdata/uvdata/tests/test_uvdata.py]: Done in 0341185

File: pyuvdata/uvdata/uvdata.py (hunk 2 above)
Comment (plaplant) [reply pyuvdata/uvdata/uvdata.py]: Done in 31c16d5

File: docs/tutorial.rst (hunk 3 above)
Comment (plaplant) [reply docs/tutorial.rst]: Done in bb9c09e

**Comment by plaplant** [comment]
@bhazelton thanks for reviewing this! I think I've addressed all of the comments, but let me know if there's anything else that should be done.

**Review by bhazelton — APPROVED** [review]
Looks good, thanks!

**Pull request merged**